### DISCUSSION TITLE
The return of DETACHED

### DISCUSSION THREAD

**Discussion by elprans** [geldata/gel#5279] [Team Posts]
Recent changes to the semantic model and the interpretation of view declarations necessitate the return of the `DETACHED` keyword.

To recap, view declarations in the `WITH` clause are no longer interpreted as a computation of a set in a nested scope. Instead, we define the turnstile operator (`:=`) as the "lexical equivalence" operator.  The `Ident := Expr` means "treat every occurrence of `Ident` as `Expr` literally, but interpret `Expr` using the context captured at the site of the declaration".

Essentially, views become a pure refactoring tool, and so the following query

```
WITH I := (SELECT Issue FILTER Issue.priority = 'High')
SELECT I
```

is perfectly equivalent to

```
(SELECT Issue FILTER Issue.priority = 'High')
```

This new interpretation is vital to the usefulness of views, as we now have the "*nested scope must not affect outer scopes*" rule, so the old "nested scope" views would be unusable in many cases.

However, the original "nested scope" interpretation was necessary to support the "detached set" case:

```
WITH I := Issue SELECT (I, Issue)  # Produced a Cartesian product of the Issue set with itself.
```

Now, `WITH I := Issue SELECT (I, Issue)` is perfectly equivalent to `SELECT (Issue, Issue)`, which, obviously, would not result in a Cartesian product.  This is because both the view declaration and the `SELECT` body are _in the same scope_.

So, basically, we need the `DETACHED` keyword back:

```
WITH I := DETACHED Issue SELECT (I, Issue)
```

The general form is `DETACHED Expr`, which means "*interpret `Expr` in a completely detached scope*".  This is equivalent to how schema-level link computables and schema-level views are interpreted.

**Comment by elprans** [comment]
`ISOLATED` may be a better candidate for the name.

**Comment by elprans** [comment]
I also happen to _like_ the explicitness of `DETACHED`.  No matter how I refactor the query, my view expression is visibly _guaranteed_ to not interact with any other scope.

**Comment by vpetrovykh** [comment]
`OTHER` may be a better alternative. It has the advantage to be clearer in the typical simple use-cases:
* `U := OTHER User`
* `(User.name, OTHER User.name)`

**Comment by vpetrovykh** [comment]
`DETACHED` seems to create an ad hoc schema level view. In case of being used in the `WITH` block, first the ad hoc view is created and then the lexical substitution takes place. Consider the following:

```
WITH MODULE test
   U := DETACHED User
SELECT User
FILTER User.name = U.name AND User != U;
```

It is important that the `U` refers to the *same* `DETAHED User` for the semantics of the query. So to keep things generalized and simple we can say that `DETACHED` "creates" a new view and "replaces" the symbols first and then `:=` aliases are all lexically substituted. This would keep `DETACHED` doing the same thing in all cases and aliasing also doing the same in all cases, regardless of what is the aliased expression.
